Model heuristics added

yafimvo opened the pull request:
## Describe your changes
This is the 1st version of the heuristics issue (#114)

## The code
In order to keep it simple, consistent, and easy to maintain I separated the code

1. We expose 2 methods `evaluate_model` and `compare_models`.
2. We use `ModelHeuristics` as a base class for different "model heuristics" we want to implement, i.e: `ModelEvaluator` and `ModelComparer`.
3. `ModelEvaluator` - evaluates specific parts of a given model using its methods
4. `ModelComparer` - compares "model a" with "model b"

## Examples

### [Report : Evaluation](https://sklearn-evaluation--232.org.readthedocs.build/en/232/classification/report_evaluation.html)

### [Report: Comparison](https://sklearn-evaluation--232.org.readthedocs.build/en/232/classification/report_comparison.html)

I also added the HTML reports (looking better than in our docs)


<!-- readthedocs-preview sklearn-evaluation start -->
----
:books: Documentation preview :books:: https://sklearn-evaluation--232.org.readthedocs.build/en/232/

<!-- readthedocs-preview sklearn-evaluation end -->

yafimvo commented:
@idomic @edublancas please share your thoughts on this

yafimvo commented:
@idomic 

> both are `evaluate_model`?

Oh it's a typo. should be `evaluate_model` and `compare_models`. fixed it.

> > It consists of different keys such as: "accuracy", "balance", "auc" etc...
> 
> I assume in that case it's only classification?

In this example - yes

Since `evaluation_state` is a dictionary we can set different keys as "evaluation parameters" and use them when generating a report.

e.g
```python
evaluation_state["cosine_similarity"]
```

edublancas commented:
Awesome! This is right on track.

My only suggestion is to embed (if possible) the generated report in the user guide.

I think having this will work:

````
from IPython.display import HTML

HTML(filename="report.html")
````

if not, we might wanna try with `IPython.display.IFrame`

idomic commented on docs/comparison/model_evaluation.md:
````diff
@@ -0,0 +1,58 @@
+```
+
+```{code-cell} ipython3
+  # download dataset. Reference: https://www.kaggle.com/datasets/redwankarimsony/heart-disease-data
````
The code in this and the next cell is over indented.

idomic commented on tests/test_models.py:
```diff
@@ -0,0 +1,41 @@
+from sklearn.model_selection import train_test_split
+
+
+def test_evaluate_model(capsys):
```
What about the rest of the functions?

idomic commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,339 @@
+    ModelEvaluator.evaluate_accuracy(y_true, y_pred, evaluation_state)
+
+    # confusion matrix
+    # ModelEvaluator.evaluate_confusion_matrix(y_true, y_pred, evaluation_state)
```
Remove unnecessary comments.

idomic reviewed (changes requested):
Looks good,
1. Are there any insights we can put in there? Maybe right before the confusion matrix?
2. I tried running it with a perfect model and it didn't generate the report (code below).
3. Agree on embedding the HTML report.

Code:

```python
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
from sklearn_evaluation import plot, table
# Based on https://github.com/Adeyinka-hub/Machine-Learning-2/blob/master/Penguin%20Dataset.ipynb

df = sns.load_dataset('penguins')

df.isnull().sum()
df.dropna(inplace=True)
Y = df.species
Y = Y.map({'Adelie': 0, 'Chinstrap':1, 'Gentoo':2})
df.drop('species', inplace=True, axis=1)
se=pd.get_dummies(df['sex'], drop_first=True) 
df = pd.concat([df, se], axis=1)
df.drop('sex', axis=1, inplace=True)
le = LabelEncoder()
df['island']= le.fit_transform(df['island']) 

X = df
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3, random_state=40)


model = RandomForestClassifier()
dt_model = model.fit(X_train, y_train)
y_pred_dt = dt_model.predict(X_test)

print("Acc on test data: {:,.3f}".format(model.score(X_test, y_test)))

model.fit(X_train, y_train)
y_pred = model.predict(X_test)
y_score = model.predict_proba(X_test)
```

yafimvo commented:
@idomic @edublancas please note that I changed the description in this PR, added docs examples, and attached example files.

I think we need to:
1. Add relevant references to our blog posts (I didn't find them)
2. Maybe add more stats to make the reports even more helpful (it's easy)
3. Go over the content and see if it makes sense
4. @idomic can you please rerun your example and see if it works? I wasn't able to reproduce it.
5. See if examples are ok (maybe a different dataset with better examples?)

edublancas commented:
Some feedback:

i see that you're hiding the output when calling  `evaluate_model`, and `compare_models` since those calls print all the generated plots. ideally, call to these functions should not print the plots so we should suppress it internally. let me take a quick look at this, I remember doing something like this before.

---

we just merged `plot.PrecisionRecall`, please rebase and include the combined plot in the report (plot1 + plot2)

---

Since `evaluate_model`, and `compare_models` and returning a `Report` object. In the tutorials, you can change this:

```python
from IPython.display import HTML
HTML(report._repr_html_())
```

for this:

```python
report
```

Jupyter will detect that `Report` has a `_repr_html_` method and use it automatically.

---

Something's broken when displaying the report (I guess that the output HTML string is malformed):

<img width="634" alt="image" src="https://user-images.githubusercontent.com/989250/215498105-340d0735-4e77-4511-a928-f4b87f23c50d.png">

---

> Model A compute time is 0.012612581253051758 

what is this compute time? is it time to produce the predictions? if so, it'd be better to say so. cause compute time is ambiguous (can be confused with time to train the model)


> Add relevant references to our blog posts (I didn't find them):

we don't have them. opened https://github.com/ploomber/sklearn-evaluation/issues/257 - feel free to comment if we have anything else missing

> Maybe add more stats to make the reports even more helpful (it's easy)

what stats are you thinking of?

edublancas commented:
> i see that you're hiding the output when calling evaluate_model, and compare_models since those calls print all the generated plots. ideally, call to these functions should not print the plots so we should suppress it internally. 

found how to hide the plots:

```python
import matplotlib.pyplot as plt
from contextlib import contextmanager

@contextmanager
def no_display_plots():
    plt.ioff()
    
    try:
        yield
    finally:
        plt.ion()

def plot():
    with no_display_plots():
        # plots inside the context manager wont be displayed
        _, ax = plt.subplots()
        ax.plot([1,1,1])
        # figures must be closed
        plt.close(ax.figure)
        
        # or to close all figures
        plt.close("all")

    return ax
```

yafimvo commented:
@edublancas this is ready for a review

edublancas commented on docs/comparison/model_comparison.md:
```diff
@@ -0,0 +1,97 @@
+---
+
+# Model comparison
+
```
~~~suggestion
```{versionadded} 0.11.2
```
~~~

edublancas commented on docs/comparison/model_comparison.md:
```diff
@@ -0,0 +1,97 @@
+
++++
+
+In this tutorial we will demonstrate how to quickly train, test and compare two models in order to find which one performs better.
```
```suggestion
In this tutorial, we will demonstrate how to quickly train, test and compare two models to find which performs better.
```

edublancas commented on docs/comparison/model_comparison.md:
````diff
@@ -0,0 +1,97 @@
+
+```
+
+## Model A
````
```suggestion
## Random forest
```

edublancas commented on docs/comparison/model_comparison.md:
````diff
@@ -0,0 +1,97 @@
+model_a.fit(X_train, y_train)
+```
+
+## Model B
````
```suggestion
## Decision tree
```

edublancas commented on docs/comparison/model_comparison.md:
```diff
@@ -0,0 +1,97 @@
+
+## Compare models
+
+Now, let's use the `compare_models` function and generate our report.
```
let's link compare_models to its API docs. I think there aren't API docs so we'd need to add them

edublancas commented on docs/comparison/model_evaluation.md:
```diff
@@ -0,0 +1,76 @@
+---
+
+# Model evaluation
+
```
~~~suggestion
```{versionadded} 0.11.2
```
~~~

edublancas commented on docs/comparison/model_evaluation.md:
```diff
@@ -0,0 +1,76 @@
+
+We use different metrics to estimate a machine learning model's performance, and to understand its strengths and weaknesses.
+
+In this guide, we'll show you how to easily generate a report with everything your need in one place using our `evaluate_model`.
```
link evaluate_model to its API docs

edublancas commented on src/sklearn_evaluation/model_heuristics/model_heuristics.py:
```diff
@@ -0,0 +1,147 @@
+import abc
+
+
+class ModelHeuristics(abc.ABC):
```
we have a `report` module, let's move this there so the imports are consistent

`from sklearn_evaluatio.report import ...`

edublancas commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,430 @@
+
+class ModelEvaluator(ModelHeuristics):
+    """
+    Generates model evaluation report
```
let's briefly document what are the rules implemented so the users know what to expect. e.g.:

- If minority class is < 5%, then a class imbalance warning is displayed
- if ... then ...

edublancas commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,430 @@
+        Checks how many of a classifier's predictions were correct,
+        and when incorrect, where the classifier got confused
+        """
+        # TODO: Implement
```
is this TODO accurate? what's missing?

edublancas commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,430 @@
+        # normalize
+        cm = cm.astype("float") / cm.sum(axis=1)[:, np.newaxis]
+
+        # TODO: Evaluate model by confusion matrix
```
what's missing?

yafimvo replied:
I found an article on evaluating random forest model, they describe how to evaluate the model using the diagonal values. I didn't implement it since I'm not sure we need it.
https://medium.com/analytics-vidhya/evaluating-a-random-forest-model-9d165595ad56

I'll remove it from the code.

edublancas commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,430 @@
+
+            # todo: better check
+            label = roc.label[i] if len(roc.label) > 0 else f"class {i}"
+            r = re.match(r"^\(class (.)*\)", label)
```
let's open an issue on this one so we don't forget to tackle it, looks important

edublancas commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,430 @@
+class ModelComparer(ModelHeuristics):
+    """
+    Compares models and generate report
+    """
```
let's document the heuristics

edublancas commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,430 @@
+
+            percision_recall_section.append_guideline(pr_a.ax_)
+
+        except Exception as exc:
```
Putting a generic `Exception` isn't recommended; it's better to put the specific exception we want to cover and only have a single line in the try body. what could it fail here? I'm guessing the model might not have predict_proba?

the same comment applies to other `try catch` statements below

edublancas commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,430 @@
+            )
+
+
+def evaluate_model(y_true, y_pred, model, y_score=None):
```
missing docstring

edublancas commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,430 @@
+    return report
+
+
+def compare_models(model_a, model_b, X_train, X_test, y_true):
```
missing docstring

edublancas commented on src/sklearn_evaluation/models.py:
```diff
@@ -0,0 +1,430 @@
+    return report
+
+
+def _check_model(model) -> None:
+    """
+    Validate if model supported
+
+    Raises
+    ~~~~~~
+    ModelNotSupported or ValueError?
+    """
+    # TODO: Implement
+    pass
+
+
+def _check_inputs(y_true, y_pred) -> None:
+    """
+    Validate if inputs supported
+
+    Raises
+    ~~~~~~
+    ModelNotSupported or ValueError?
+    """
+    # TODO: Implement
+    # TODO: If optional args given test them
+    pass
```
delete if they wont be implemented

edublancas commented on tests/test_models.py:
```diff
@@ -0,0 +1,75 @@
+
+@pytest.fixture
+def heart_dataset():
+    import urllib.request
+    import pandas as pd
```
move to the top of the file

edublancas commented on tests/test_models.py:
```diff
@@ -0,0 +1,75 @@
+    _clean_files([file_name, "example-compare-report.html"])
+
+
+def _clean_files(files):
```
you can use a fixture to run the test in a tmp directory so you don't have to clean up the files. this ensures that the files are deleted even if the test fails.

the fixture is called `tmp_directory`

edublancas commented on tests/test_models.py:
```diff
@@ -0,0 +1,75 @@
+    return data, file_name
+
+
+def test_evaluate_model(heart_dataset):
```
```suggestion
def test_evaluate_model(heart_dataset, tmp_directory):
```

edublancas commented on tests/test_models.py:
```diff
@@ -0,0 +1,75 @@
+    report = evaluate_model(y_test, y_pred, model=model, y_score=y_score)  # noqa
+    report.save("example-report.html")
+
+    _clean_files([file_name, "example-report.html"])
```
```suggestion
```

edublancas commented on tests/test_models.py:
```diff
@@ -0,0 +1,75 @@
+    _clean_files([file_name, "example-report.html"])
+
+
+def test_compare_models(heart_dataset):
```
```suggestion
def test_compare_models(heart_dataset, tmp_directory):
```

edublancas commented on tests/test_models.py:
```diff
@@ -0,0 +1,75 @@
+    report = compare_models(model_a, model_b, X_train, X_test, y_test)  # noqa
+    report.save("example-compare-report.html")
+
+    _clean_files([file_name, "example-compare-report.html"])
```
```suggestion
```

edublancas commented on tests/test_models.py:
```diff
@@ -0,0 +1,75 @@
+    _clean_files([file_name, "example-compare-report.html"])
+
+
+def _clean_files(files):
+    for file in files:
+        pathlib.Path(file).unlink()
```
```suggestion
```

yafimvo commented:
@edublancas I added [report-evaluation](https://sklearn-evaluation--232.org.readthedocs.build/en/232/api/plot.html#report-evaluation) and [report-comparison](https://sklearn-evaluation--232.org.readthedocs.build/en/232/api/plot.html#report-comparison) to the API section but it only shows the plots without the HTML. any suggestions?

edublancas commented:
> I added [report-evaluation](https://sklearn-evaluation--232.org.readthedocs.build/en/232/api/plot.html#report-evaluation) and [report-comparison](https://sklearn-evaluation--232.org.readthedocs.build/en/232/api/plot.html#report-comparison) to the API section but it only shows the plots without the HTML. any suggestions?

they reports won't display, this is a limitation of the directive we're using to document the examples, but that's fine. Let's just link from the docstring to the tutorial to ensure people can see the sample reports.

---

is this ready for review? I saw you requested a review but noticed that you didn't reply to some of my comments from the  latest review, so I'm unsure if you addressed the or if they're still pending.

yafimvo commented:
@edublancas yes. for some reason it didn't catch the resolve. I added more tests.

I think besides the code, please check the relevance of the reports. If something is missing or the calculation is inaccurate this can be easily fixed.

edublancas commented on examples/heart.csv:
```diff
@@ -0,0 +1,304 @@
+﻿age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
```
I think this was included by mistake since we are downloading it in the tutorial?

yafimvo replied:
yes. removed it.

edublancas commented on docs/api/plot.rst:
```diff
@@ -152,3 +152,11 @@ validation_curve
+
+report_evaluation
+-----------------
+.. autofunction:: sklearn_evaluation.models.evaluate_model
```
maybe I'm not recalling correctly, but I think we decided to move this to the report module?

so `from sklearn_evaluation.report import evaluate_model`?

yafimvo replied:
yes. changed it.

edublancas commented:
hey @neelasha23: can you take a look? I already reviewed the code but we'd like some user testing:

> I think besides the code, please check the relevance of the reports. If something is missing or the calculation is inaccurate this can be easily fixed.


essentially, generate the reports with different models, input datasets and see if they make sense

neelasha23 commented:
There are some issues in the Precision Recall curve:

Line 254 we need to add check for `str` type :
```
        if n_classes > 2:
            if label is not None:
                if isinstance(label, str):
                    label = [label]
                label = ["Micro-average"] + label
```

But even after adding this I get this error: 
```
ValueError: Found input variables with inconsistent numbers of samples: [226280, 150640]
If you need help solving this issue, send us a message: https://ploomber.io/community
```

PFB the code
[model heuristics.txt](https://github.com/ploomber/sklearn-evaluation/files/10821172/model.heuristics.txt)

1. Is there a way to ignore a plot during report generation?
2. Can we add debug logs here?

neelasha23 commented on docs/classification/report_evaluation.md:
````diff
@@ -0,0 +1,80 @@
+```
+
+## Display the report
+
````
Here also we can add how to save the report as html

neelasha23 commented:
For the same dataset shared above the evaluate report looks like this:
1. There isn't much insights in the AUC value , probably because this is a multi-class classification? Is this expected?
2. Do we mean `dataset is balanced` instead of `model is balanced`? A short explanation here would be helpful

<img width="1337" alt="Screenshot 2023-02-24 at 12 07 25 PM" src="https://user-images.githubusercontent.com/29096135/221109416-9043d762-64b7-4c3b-94ab-19aacca2681d.png">

<img width="1292" alt="Screenshot 2023-02-24 at 12 04 27 PM" src="https://user-images.githubusercontent.com/29096135/221109288-97bc495b-fb6a-4362-baa6-79e89f114ae7.png">

neelasha23 commented:
Facing similar issues with the `mushrooms` dataset (https://www.kaggle.com/datasets/uciml/mushroom-classification)

Code:

```
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
data = pd.read_csv("/Users/neelashasen/Downloads/mushrooms.csv")
from sklearn.preprocessing import LabelEncoder
labelencoder=LabelEncoder()
for col in data.columns:
    data[col] = labelencoder.fit_transform(data[col])
X = data.iloc[:,1:23]
y = data.iloc[:, 0]
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X=scaler.fit_transform(X)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=4)
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn import metrics

model_LR= LogisticRegression()
model_LR.fit(X_train,y_train)
y_prob = model_LR.predict_proba(X_test)[:,1] 
y_pred = np.where(y_prob > 0.5, 1, 0) 
model_LR.score(X_test, y_pred)
y_score = model_LR.predict_proba(X_test)
from sklearn_evaluation.models import evaluate_model

report = evaluate_model(model_LR, y_test, y_pred, y_score=y_score)
```

<img width="1090" alt="Screenshot 2023-02-24 at 12 50 39 PM" src="https://user-images.githubusercontent.com/29096135/221116878-a5820076-f3ab-4fd0-a365-725e4b281d82.png">

yafimvo commented:
@neelasha23 

> * Is there a way to ignore a plot during report generation?
> * Can we add debug logs here?

I added an error and stack trace to the report. Also, if something fails - it now captures any exception and logs it to the report.

> 1. There isn't much insights in the AUC value , probably because this is a multi-class classification? Is this expected?
> 2. Do we mean `dataset is balanced` instead of `model is balanced`? A short explanation here would be helpful
> 3. I think in `General Stats` we can also include some common metrics like `Precision`, `Recall` etc
> 4. Title of the report is `Model evaluation - Pipeline` .. can users pass a parameter to change this name?
> 5. Can we add some insights as to which classes performed the best and which classes performed the worst?

1. Yes. If everything is OK I didn't show the AUC. changed it.
2. I'm not sure which term is correct. Please let me know which one you think is more suitable here.
3. Added Precision and recall plot.
4. We can now change the title by passing `evaluate_model(model, ..., report_title = "custom title" )` or `compare_models(..., report_title = "custom title" )`
5. Added top AUC, worst AUC and number of classes.

neelasha23 commented:
> > Line 254 we need to add check for `str` type :
> 
> @edublancas Do you think we should change it here or open a new issue?

On second thoughts, I think this change is not needed. Because when `n_classes > 2` label should be a list.
We need to check why the evaluator is passing a string here as label for multiclass classification.

@yafimvo

idomic commented:
> > Line 254 we need to add check for `str` type :
> 
> @edublancas Do you think we should change it here or open a new issue?

I think as a role of thumb if it's not blocking and the pr is open for more than 2 weeks, yes we should create a new issue and a PR just for this small fix.

neelasha23 commented:
> 2. I'm not sure which term is correct. Please let me know which one you think is more suitable here.

Please change to `dataset is balanced`.

@yafimvo

edublancas commented:
fixes the docs in master, please rebase
